### PR TITLE
storageparams: Introduce "schema_locked" which prevents schema changes

### LINKED ISSUE #100766: add new storage parameter to disable schema changes (DDLs) for a particular table
We need to add a new storage parameter called `disable_schema_changes` that can be set at the table table as described in RFC [#99890](https://github.com/cockroachdb/cockroach/pull/99890) for supporting reduced latency changefeeds. Setting this should disable any DDLs that are run against the table. DCLs (GRANT, COMMENT) should not be affected and should be allowed.

Setting:
```
ALTER TABLE t SET (disable_schema_changes=t);
```

Unsetting:
```
ALTER TABLE t SET (disable_schema_changes=f);
```
Trying to run a DDL on a table with the storage parameter should fail with an error along the lines of:
```
The storage parameter disable_schema_changes is set to true on this table, please set it to false by running 
ALTER TABLE t SET (disable_schema_changes=f); before doing a schema change.
```

Further `disable_schema_changes` should only be allowed to be set and unset by the owner of that particular table.

Jira issue: CRDB-26611

Epic CRDB-2258

### PR DESCRIPTION
**Pull request by Xiang-Gu**
We introduced a new table storage parameter "schema_locked", which, when set, prevents schema changes on that table. This parameter is only allowed to set/reset in a single-statement implicit transaction.
Note that we whitelisted certain schema change stmts (e.g. GRANT, COMMENT) so that they are allowed even if this parameter is set.

Fixes #100766
Epic: CRDB-2258
Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
@cockroachdb/sql-queries could someone volunteer to review this on behalf of Queries team?

**Review by rafiss — CHANGES_REQUESTED** [review]
nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-:-NT2e7uREEii7tPsfA7Z:b-j4xmku)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[`pkg/sql/alter_index.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2b5NO0uafZUbPW7Kg:-NT2b5NO0uafZUbPW7Kh:b6b83fe) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/alter_index.go#L53)):*
> ```Go
> 
> 	// Disallow schema changes if this table's schema is locked.
> 	if err = checkTableSchemaUnlocked(tableDesc); err != nil {
> ```

nit: when using the pattern of assigning the error in the `if` statement, it's more conventional to use `if err := ...; err != nil` so that this err stays in the scope of the if statement. (same for below changes)
___
*[`pkg/sql/pgwire/pgcode/codes.go` line 375 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2cqEK9U-XskxdGBmv:-NT2cqEK9U-XskxdGBmw:brncurb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/pgwire/pgcode/codes.go#L375)):*
> ```Go
> 	// SchemaChangeOnSchemaLockedTable signals that a DDL is attempted on a table
> 	// whose schema is locked.
> 	SchemaChangeOnSchemaLockedTable = MakeCode("55C03")
> ```

nit: custom error codes should be defined under `Class XC - cockroach extension.` (see below in this file)

but in this case, we could consider reusing the existing code `OperatorIntervention`. though that may not make sense if the schema locking happens automatically, and not at the request of a human operator.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_table_schema_locked.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2bif56qn-a4TXbnRb:-NT2bif56qn-a4TXbnRc:b-bs8ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/schemachanger/scplan/internal/opgen/opgen_table_schema_locked.go#L18)):*
> ```Go
> )
> 
> func init() {
> ```

admittedly i'm new to this code, but it feels like some of this could have comments.
___
*[`pkg/sql/sqlerrors/errors.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2cLZoCN_dRcH-EBnY:-NT2cLZoCN_dRcH-EBnZ:b-36uc0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/sqlerrors/errors.go#L42)):*
> ```Go
> func NewSchemaChangeToTableWithLockedSchema(tableName string) error {
> 	return errors.WithHintf(pgerror.Newf(pgcode.SchemaChangeOnSchemaLockedTable,
> 		`schema changes are disallowed on table %q because its schema is locked.`, tableName),
> ```

nit: i think "its schema is locked" can be confusing. people may think it means the schema in which the table is defined. maybe say `schema changes are disallowed on table %q because it is locked` (note no trailing period)

also the function can be called `NewSchemaChangeToLockedTableError`
___
*[`pkg/sql/storageparam/storage_param.go` line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2cBsM91lmrhr7FnNO:-NT2cBsN9VSRMSmCC-VC:b7j5ru6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/storageparam/storage_param.go#L148)):*
> ```Go
> 			// change we make to the descriptor in the transaction, so we can uphold
> 			// the "one-version invariant" as discussed further in RFC
> 			// https://github.com/ajwerner/cockroach/blob/ajwerner/low-latency-rfc-take-3/docs/RFCS/20230328_low_latency_changefeeds.md
> ```

thanks for this good explanation!

**Review by Xiang-Gu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/alter_index.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2b5NO0uafZUbPW7Kg:-NTJcgIF6hnAWptAqZ-X:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/alter_index.go#L53)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: when using the pattern of assigning the error in the `if` statement, it's more conventional to use `if err := ...; err != nil` so that this err stays in the scope of the if statement. (same for below changes)
</blockquote></details>

done



_Code quote:_
```Go
/ Disallow schema changes if this table's schema is locke
```
___
*[`pkg/sql/pgwire/pgcode/codes.go` line 375 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2cqEK9U-XskxdGBmv:-NTJdAcb5edc1Ow2et-g:b3d35so) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/pgwire/pgcode/codes.go#L375)):*
> but in this case, we could consider reusing the existing code OperatorIntervention. though that may not make sense if the schema locking happens automatically, and not at the request of a human operator.

I don't have strong feelings here; I think in the foreseeable future we will not make this schema locking/unlocking happen automatically so I will use `OperatorIntervention`.

> nit: custom error codes should be defined under Class XC - cockroach extension. (see below in this file)

I think `55Cxx` is also a Cockroach extension for errors category "Object Not In Prerequisite State" and that was why I put it there.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_table_schema_locked.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2bif56qn-a4TXbnRb:-NTJe6tFAH9YfU-byEKq:b-7rl0kd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/schemachanger/scplan/internal/opgen/opgen_table_schema_locked.go#L18)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

admittedly i'm new to this code, but it feels like some of this could have comments.
</blockquote></details>

done
___
*[`pkg/sql/sqlerrors/errors.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2cLZoCN_dRcH-EBnY:-NTJfPBfAVe-Qhs8mR_5:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/sqlerrors/errors.go#L42)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: i think "its schema is locked" can be confusing. people may think it means the schema in which the table is defined. maybe say `schema changes are disallowed on table %q because it is locked` (note no trailing period)

also the function can be called `NewSchemaChangeToLockedTableError`
</blockquote></details>

done



_Code quote:_
```Go
NewSchemaChangeToTableWithLockedSchem
```
___
*[`pkg/sql/storageparam/storage_param.go` line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NT2cBsM91lmrhr7FnNO:-NTJg2y37djjX9DHGeRg:b32vhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573134e933c76986be1aa5ff8cc09bd537c8234/pkg/sql/storageparam/storage_param.go#L148)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

thanks for this good explanation! 
</blockquote></details>

:)

**Review by miretskiy — COMMENTED** [review]
> Reviewed 17 of 45 files at r1, 12 of 15 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @Xiang-Gu)

___
*[`pkg/sql/sqlerrors/errors.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTL37rg3X9FJ1fyOk84:-NTL37rg3X9FJ1fyOk85:bsl4b59) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c566800f4e2bd3386b315f850df245f2b9af98d/pkg/sql/sqlerrors/errors.go#L40)):*
> ```Go
> // NewSchemaChangeToLockedTableErr creates an error signaling schema
> // change statement is attempted on a table with locked schema.
> func NewSchemaChangeToLockedTableErr(tableName string) error {
> ```

nits/question: Is it always a table? Do we plan on eventually having locks on other types of objects?
should "ToLocked" be changed to "OnLocked"?
___
*[`pkg/sql/sqlerrors/errors.go` line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTL3vQT1cI-JOjQmlP4:-NTL3vQT1cI-JOjQmlP5:b-6u4g1t) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c566800f4e2bd3386b315f850df245f2b9af98d/pkg/sql/sqlerrors/errors.go#L43)):*
> ```Go
> 	return errors.WithHintf(pgerror.Newf(pgcode.OperatorIntervention,
> 		`schema changes are disallowed on table %q because it is locked`, tableName),
> 		"To unlock the table, try ALTER TABLE %v SET (schema_locked = false);", tableName)
> ```

Perhaps expand this error message to indicate that ALTER TABLE ... (schema_locked = true) *after* schema change completes, might be recommended.

(Is there no way to do this automatically?). Or is it something to do in the future?
___
*[`pkg/sql/storageparam/storage_param.go` line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTL3cmJ3INT8-iNkHeA:-NTL3cmJ3INT8-iNkHeB:b-gmfxl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c566800f4e2bd3386b315f850df245f2b9af98d/pkg/sql/storageparam/storage_param.go#L151)):*
> ```Go
> 			if len(keys) > 1 || !evalCtx.TxnImplicit || !evalCtx.TxnIsSingleStmt {
> 				return pgerror.Newf(pgcode.InvalidParameterValue, "%q can only be set/reset on "+
> 					"its own without other parameters in a single-statement implicit transaction.", key)
> ```

The operator is responsible for flipping this bit back to locked manually after schema change completes?
That's fine; but perhaps mention this fact in this error message.

**Review by miretskiy — COMMENTED** [review]
Super nice work; I'm not an arbiter of schema changes, so my comments are fyi/nice to have.  Feel free to ignore.  Or not :)
Either way, super exciting.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-:-NTL4fD30NQ-sSVTnwyS:b-jkuusv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @Xiang-Gu)

**Review by Xiang-Gu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @rafiss)

___
*[`pkg/sql/sqlerrors/errors.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTL37rg3X9FJ1fyOk84:-NTP9Puj7ibgC-NL4Wga:bu1p38p) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c566800f4e2bd3386b315f850df245f2b9af98d/pkg/sql/sqlerrors/errors.go#L40)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nits/question: Is it always a table? Do we plan on eventually having locks on other types of objects?
should "ToLocked" be changed to "OnLocked"?
</blockquote></details>

In principle, we can have locks on either tables or types, but the choice to make it a storage parameter implies it will always be a table (types do not have storage parameters).


Changed to `OnLocked`.
___
*[`pkg/sql/sqlerrors/errors.go` line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTL3vQT1cI-JOjQmlP4:-NTPBaRF0R0xcjWAPdxs:b-mcsuu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c566800f4e2bd3386b315f850df245f2b9af98d/pkg/sql/sqlerrors/errors.go#L43)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Perhaps expand this error message to indicate that ALTER TABLE ... (schema_locked = true) *after* schema change completes, might be recommended.

(Is there no way to do this automatically?). Or is it something to do in the future?
</blockquote></details>

done.

It can be done in the declarative schema changer in the future (phase 2 of the RFC)
___
*[`pkg/sql/storageparam/storage_param.go` line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTL3cmJ3INT8-iNkHeA:-NTPBpCW0L7tGSwgN_H9:b770sdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c566800f4e2bd3386b315f850df245f2b9af98d/pkg/sql/storageparam/storage_param.go#L151)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

The operator is responsible for flipping this bit back to locked manually after schema change completes?
That's fine; but perhaps mention this fact in this error message.
</blockquote></details>

That's done in the error message above. This error msg is for a different purpose -- it ensures we only set/reset this locked-bit in a single-statement implicit transaction.



_Code quote:_
```Go
its own without other parameters in
```

**Review by miretskiy — COMMENTED** [review]
> Reviewed 1 of 14 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

**Comment by Xiang-Gu** [comment]
We are waiting for the decision as to whether or not we want to backport this to v23.1 (so we can later backport https://github.com/cockroachdb/cockroach/pull/101694 to v23.1). This will affect on what cluster version do we gate set/reset this new storage parameter.

**Review by fqazi — APPROVED** [review]
> Reviewed 20 of 45 files at r1, 10 of 15 files at r2, 5 of 14 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @rafiss, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/screl/scalars.go` line 133 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtT2heE9jOezDEyNe-:-NTtT2heE9jOezDEyNe0:b-rbitsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0b5f352a3ae195b57e2cdf63d4844fb7bfe2d63/pkg/sql/schemachanger/screl/scalars.go#L133)):*
> ```Go
> 	case *scpb.TableSchemaLocked:
> 		// TODO (xiang): When `V23_2` is created, change it to `V23_2`.
> 		return clusterversion.V23_2Start
> ```

We should get an issue open for this, I'm super paranoid about having broken version gates. It can be a meta issue that says we should replace all 23_2 starts before 23.2 cuts
___
*[`pkg/sql/logictest/testdata/logic_test/schema_locked` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtTbR9DhHG7inmQvJn:-NTtTbR9DhHG7inmQvJo:bxaj6u8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0b5f352a3ae195b57e2cdf63d4844fb7bfe2d63/pkg/sql/logictest/testdata/logic_test/schema_locked#L2)):*
> ```
> statement ok
> CREATE TABLE t (i INT PRIMARY KEY) WITH (schema_locked = t);
> ```

Address this later, but we should add mixed version coverage to ensure it's blocked until the entire cluster understands.

**Review by rafiss — APPROVED** [review]
i only have nits about comments, and the decision to use V23_1 as the gate. lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-:-NTtWKf8CTCknno9hKf-:b-ydy3lv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, and @Xiang-Gu)

___
*[`-- commits` line 4 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtUYjz34IXDjcAm2VJ):*
let's keep the commit message, but remove the release note. we will keep this for internal usage only.
___
*[`pkg/sql/schemachanger/screl/scalars.go` line 133 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtT2heE9jOezDEyNe-:-NTtVSSvErEgP3w3WUu_:b3fko5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0b5f352a3ae195b57e2cdf63d4844fb7bfe2d63/pkg/sql/schemachanger/screl/scalars.go#L133)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

We should get an issue open for this, I'm super paranoid about having broken version gates. It can be a meta issue that says we should replace all 23_2 starts before 23.2 cuts
</blockquote></details>

as per our slack discussion, let's just go ahead and make this `V23_1`
___
*[`pkg/sql/sqlerrors/errors.go` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtVhZW4M1prAhMAed5:-NTtVhZW4M1prAhMAed6:b-rjuy11) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0b5f352a3ae195b57e2cdf63d4844fb7bfe2d63/pkg/sql/sqlerrors/errors.go#L43)):*
> ```Go
> 	return errors.WithHintf(pgerror.Newf(pgcode.OperatorIntervention,
> 		`schema changes are disallowed on table %q because it is locked`, tableName),
> 		"To unlock the table, try \"ALTER TABLE %v SET (schema_locked = false);\" "+
> ```

we should remove this hint, since we don't want users to use this directly. instead, we can just keep the main error message and perhaps have it say "schema changes are disallowed on table %q because it is locked by an internal operation"

File: pkg/sql/alter_table_set_schema.go [hunk 1]
```diff
@@ -88,6 +88,11 @@ func (p *planner) AlterTableSetSchema(
 		}
 	}
 
+	// Disallow schema changes if this table's schema is locked.
+	if err := checkTableSchemaUnlocked(tableDesc); err != nil {
```
Comment (chengxiong-ruan) [inline pkg/sql/alter_table_set_schema.go:92]: Do we need a check also in `alter_table_owner.go`?

**Review by chengxiong-ruan — APPROVED** [review]
LGTM! nice work.

**Review by Xiang-Gu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @miretskiy, and @rafiss)

___
*[`-- commits` line 4 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtUYjz34IXDjcAm2VJ):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

let's keep the commit message, but remove the release note. we will keep this for internal usage only.
</blockquote></details>

Done.
___
*[`pkg/sql/alter_table_set_schema.go` line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#gh-1176851324:-NTtaXYB7qr6e6p3k1vE:buqxji1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0b5f352a3ae195b57e2cdf63d4844fb7bfe2d63/pkg/sql/alter_table_set_schema.go#L92)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Do we need a check also in `alter_table_owner.go`?
</blockquote></details>

I made the decision not to because I didn't enforce this locked-bit for GRANT and the principle here is don't enforce it for any privilege/role/user related schema changes.
___
*[`pkg/sql/schemachanger/screl/scalars.go` line 133 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtT2heE9jOezDEyNe-:-NTtbN1o4rlWTIpEqWfc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0b5f352a3ae195b57e2cdf63d4844fb7bfe2d63/pkg/sql/schemachanger/screl/scalars.go#L133)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

as per our slack discussion, let's just go ahead and make this `V23_1`
</blockquote></details>

Done.
___
*[`pkg/sql/sqlerrors/errors.go` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtVhZW4M1prAhMAed5:-NTtbOCZChlZIYB68cpB:b-1ei5zp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0b5f352a3ae195b57e2cdf63d4844fb7bfe2d63/pkg/sql/sqlerrors/errors.go#L43)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we should remove this hint, since we don't want users to use this directly. instead, we can just keep the main error message and perhaps have it say "schema changes are disallowed on table %q because it is locked by an internal operation"
</blockquote></details>

discussed offline and we decided to keep it
___
*[`pkg/sql/logictest/testdata/logic_test/schema_locked` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtTbR9DhHG7inmQvJn:-NTtg337Et7AbnIBkq1Q:bragiyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0b5f352a3ae195b57e2cdf63d4844fb7bfe2d63/pkg/sql/logictest/testdata/logic_test/schema_locked#L2)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Address this later, but we should add mixed version coverage to ensure it's blocked until the entire cluster understands.
</blockquote></details>

done. I've created a new logictest file that will *only* run under the `local-mixed-22.2-23.1` cluster version

**Comment by Xiang-Gu** [comment]
TFTR!

bors r+

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, and @miretskiy)

___
*[`pkg/sql/schemachanger/screl/scalars.go` line 133 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101487#-NTtT2heE9jOezDEyNe-:-NTuJ0PQ4Mh6tAN_tyMD:bs6ldii) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0b5f352a3ae195b57e2cdf63d4844fb7bfe2d63/pkg/sql/schemachanger/screl/scalars.go#L133)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

btw, for other features that do need `23_2`, we can just use this alias: https://github.com/cockroachdb/cockroach/blob/1927969af281fd1a1312fc86c95a456d48e6d893/pkg/clusterversion/cockroach_versions.go#L1020

**Pull request merged**